### PR TITLE
Jhbate/timings

### PR DESCRIPTION
**Pull request by jh-bate**
@jebeck can you have a look at this - its just some simple log statements to better understand the time taken for queries for easier support

... and broke the build so will check tomorrow

### CODE REVIEW

File: clients/mongoStoreClient.go [hunk 1]
```diff
@@ -118,8 +119,8 @@ func constructQuery(details *model.QueryData) (query bson.M, sort string) {
 	for _, v := range details.MetaQuery {
 		log.Println("constructQuery: create base queries")
 		//base query
-		queryThis := bson.M{"groupId": v}
-		queryThat := bson.M{"_groupId": v, "_active": true}
+		queryThat := bson.M{"groupId": v}
+		queryThis := bson.M{"_groupId": v, "_active": true}
```
Comment (jebeck) [inline clients/mongoStoreClient.go]: What are the implications of the fact that `"_active": true` wasn't there before? Any?

**Comment by jebeck** [comment]
One question, and also the build is failing - just restarted it, but it looks legit, not a Travis hiccup. Do some tests need updating?

**Comment by jh-bate** [comment]
@darinkrauss if you could take a look at this it would be appreciated. I have added indexes based on query usage and also some basic timings and logging.

File: clients/mongoStoreClient.go [hunk 2]
```diff
@@ -45,6 +47,21 @@ func NewMongoStoreClient(config *mongo.Config) *MongoStoreClient {
...
+	}
+	mongoSession.DB("").C(DEVICE_DATA_COLLECTION).EnsureIndex(queryIndex)
+
+	//As above but includes uploadId for restriction of data returned
+	queryUploadIdIndex := mgo.Index{
+		Key:        []string{"_groupId", "_active", "type", "uploadId", "-time"},
+		Background: true,
+	}
+	mongoSession.DB("").C(DEVICE_DATA_COLLECTION).EnsureIndex(queryUploadIdIndex)
+
```
Comment (darinkrauss) [inline clients/mongoStoreClient.go]: @jh-bate Have you gotten any performance results back from adding these indices on your development box?  Or will we try this on prod or staging?

**Comment by darinkrauss** [comment]
@jh-bate Overall it looks good with my limited understanding at this time.  I'll be interested to see what performance gains you've seen with those indices on dev, staging, and prod.

**Comment by jh-bate** [comment]
Also FYI http://godoc.org/gopkg.in/mgo.v2#Collection.EnsureIndex and this leads to the mongo docs that I went through also

**Comment by darinkrauss** [comment]
@jh-bate Hmm, seems odd that devel would use a different index from local.  Are there some legacy indices on devel or indices setup on devel/staging/prod that we need setup locally?

**Comment by jh-bate** [comment]
tide-whisperer also has an index and access the same collection https://github.com/tidepool-org/tide-whisperer/blob/master/tide-whisperer.go#L112 which is for this query https://github.com/tidepool-org/tide-whisperer/blob/master/tide-whisperer.go#L154. It was getting  late last night but i wondered if there where old redundant indexes that should be cleaned up and then the new ones applied - I plan investigating later tonight

**Comment by darinkrauss** [comment]
Actually, this brings up an interesting point.  I note that the code creates collections and indexes on-the-fly.  I think we should move to a "bootstrap then fail" process instead.  Basically, a script that sets up the environment to a given state at environment creation time.  Then the API code itself should fail hard if the collection/index isn't there (because it indicates something is wrong or unexpected).  Just my 2 cents...

**Comment by jh-bate** [comment]
True - will discuss further. Essentially the test code, where I test that the required indexes do actually exist, could move into the api itself

**Comment by jh-bate** [comment]
@darinkrauss a couple of updates

- I have added gzip compression and also 
- ensured we are capturing the failure of the mongo query itself and taking the appropriate action (i.e. log the failure and return and error to the user)

**Pull request merged**